lib/fluent/plugin/kubernetes_metadata_watch_pods.rb

```diff
@@ -24,28 +24,44 @@ module WatchPods
     include ::KubernetesMetadata::Common
 
     def start_pod_watch
-      begin
-        options = {
-          resource_version: '0'  # Fetch from API server.
-        }
-        if ENV['K8S_NODE_NAME']
-          options[:field_selector] = 'spec.nodeName=' + ENV['K8S_NODE_NAME']
-        end
-        pods = @client.get_pods(options)
-        pods.each do |pod|
-          cache_key = pod.metadata['uid']
-          @cache[cache_key] = parse_pod_metadata(pod)
-          @stats.bump(:pod_cache_host_updates)
-        end
-        options[:resource_version] = pods.resourceVersion
-        watcher = @client.watch_pods(options)
-      rescue Exception => e
-        message = "Exception encountered fetching metadata from Kubernetes API endpoint: #{e.message}"
-        message += " (#{e.response})" if e.respond_to?(:response)
+      get_pods_and_start_watcher
+    rescue Exception => e
+      message = "start_pod_watch: Exception encountered setting up pod watch " \
+                "from Kubernetes API #{@apiVersion} endpoint " \
+                "#{@kubernetes_url}: #{e.message}"
+      message += " (#{e.response})" if e.respond_to?(:response)
+      log.debug(message)
+
+      raise Fluent::ConfigError, message
+    end
 
-        raise Fluent::ConfigError, message
+    # List all pods, record the resourceVersion and return a watcher starting
+    # from that resourceVersion.
+    def get_pods_and_start_watcher
+      options = {
+        resource_version: '0'  # Fetch from API server.
+      }
+      if ENV['K8S_NODE_NAME']
+        options[:field_selector] = 'spec.nodeName=' + ENV['K8S_NODE_NAME']
       end
+      pods = @client.get_pods(options)
+      pods.each do |pod|
+        cache_key = pod.metadata['uid']
+        @cache[cache_key] = parse_pod_metadata(pod)
+        @stats.bump(:pod_cache_host_updates)
+      end
+      options[:resource_version] = pods.resourceVersion
+      watcher = @client.watch_pods(options)
+      Thread.current[:watcher_alive] = true
+      watcher
+    end
 
+    # Given a watcher, process the notices and handle retries by resetting the
+    # watcher if needed.
```

Comment:
I was confused by the API at first. It seems weird that this method takes a watcher, but can also instantiate its own. It would be clearer if it only did one of those things. Do you think it can be refactored?

Reply:
Sure thing. Refactored.

Reply:
Thanks, it's clearer now. Unless I'm missing something, I think it could benefit from being even simpler. E.g.:
```ruby
wacher = nil
while thread_current_running?
  watcher ||= get_namespaces_and_start_watcher
  begin
    process_namespace_watcher_notices_with_retries(watcher)
  rescue Exception => e
    log.info "Exception encountered parsing namespace watch event. " \
             "The connection might have been closed. " \
             "Resetting the namespace watcher.", e
    watcher = nil
    sleep(@watch_retry_interval)
  end
end
```

Reply:
Refactored to move exception handling to the main filter file. Also added retry_backoff and retry_max_times as suggested in the other thread.